File: JobTracker/UserInterface/app/types/applications/jobApplication.ts
import { Posting } from "../jobs/posting";

export interface JobApplication {
  Id: number,
  JobId: number,
  Posting: Posting,
  CoverLetter: string,
  AppliedAt: Date,
  LastStatusChangeAt: Date,
  Status: ApplicationStatus
  StatusHistory: ApplicationStatusHistory[]
}

export enum ApplicationStatus {
  Pending,
  Submitted,
  Interview,
  Offer,
  Accepted,
  Rejected,
  Ghosted
}

export interface ApplicationStatusHistory {
  Id: number,
  JobApplicationId: number,
  JobApplication: JobApplication,
  ApplicationStatus: ApplicationStatus,
  ChangedAt: Date,
  Note?: string
}

File: JobTracker/Application/Features/JobApplication/JobApplication.cs
using JobTracker.Application.Features.JobSearch;

namespace JobTracker.Application.Features.JobApplication;

public class JobApplication
{
    public int Id { get; set; }
    public int ProfileId { get; set; }
    public int JobId { get; set; }
    public Posting Posting { get; set; } = null!;
    public string CoverLetter { get; set; } = null!;
    public DateTime AppliedAt { get; set; }
    public DateTime? LastStatusChangeAt { get; set; }
    public ApplicationStatus Status { get; set; }
    public List<ApplicationStatusHistory> StatusHistory { get; set; } = new();
    //public List<Mails> RelatedMails
}

public enum ApplicationStatus
{
    Pending = 0,
    Submitted = 1,
    Interview = 2,
    Offer = 3,
    Accepted = 4,
    Rejected = 5,
    Ghosted = 6
}